src/main/java/com/google/sps/servlets/UserCardsServlet.java

```diff
@@ -75,36 +83,39 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     if (userService.isUserLoggedIn()) {
 
-        // if blobKey is null, user did not add an image
         String folderKey = request.getParameter("folderKey");
-        String blobKey = getBlobKeyfromBlobstore(request, "image");
         String labels = request.getParameter("labels");
         String fromLang = request.getParameter("fromLang");
         String toLang = request.getParameter("toLang");
-        String textNotTranslated = request.getParameter("textNotTranslated");
-        String textTranslated = translateText(textNotTranslated, toLang);
+        String rawText = request.getParameter("rawText");
+        String textTranslated = request.getParameter("translatedText");
+        String blobKey = getBlobKey(request);
 
-        Card card = new Card(blobKey, labels, fromLang, toLang, textNotTranslated, textTranslated);
+        Card card = new Card(blobKey, labels, fromLang, toLang, rawText, textTranslated);
         Entity cardEntity = card.createEntity(KeyFactory.stringToKey(folderKey));
 
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         datastore.put(cardEntity);
     }
-
   }
+  
+  private String getBlobKey(HttpServletRequest request){
+    // Method to determine whether or not this is a unit test or live server
+    // Unit tests will always set blobKey to "null"
+    // There should be no paramater testStatus in the live server thus returns null
+    String blobKey;
+
+    if (request.getParameter("testStatus") == null) {
+      blobKey = getBlobKeyfromBlobstore(request, "image");
```

Comment:
Why do we need to special-case the test server? Does `getBlobKeyfromBlobstore` fail on the test server? The implementation looks like it would still work.

Reply:
More specifically, the method under getBlobKeyFromBlobstore, blobstoreService.getUploads(request) throws an error. If I could remember, it had something to do with needing to call the UserCardsServlet through a callback function, which is the method in UploadOnBlobstoreServlet --> String uploadUrl = blobstoreService.createUploadUrl("/usercards")". I didn't know how to mock the process where the user can upload an image through the created upload url. I tried looking into localBlobstore config and any particular examples on stackoverflow, but couldn't find any. So, I ended up isolating that process throughout my unit tests. 

I think the implementation would work, since I tried it in my portfolio, but I was hoping we could test it when we integrate the servlet with the front end.

Reply:
Note: Leave a comment